File: backend/app/schemas/pipeline.py
from typing import List, Optional, Literal
from datetime import datetime
from pydantic import BaseModel, ConfigDict


PIPELINE_STAGES = [
    "Applied",
    "Screening",
    "Shortlisted",
    "Interview",
    "Technical",
    "HR",
    "Offer",
    "Hired",
    "Rejected",
]


# ── Stage move ────────────────────────────────────────────────────────────────

class ApplicationStageUpdate(BaseModel):
    stage: Literal[
        "Applied", "Screening", "Shortlisted", "Interview",
        "Technical", "HR", "Offer", "Hired", "Rejected"
    ]
    pipeline_order: Optional[int] = None


class ApplicationCreate(BaseModel):
    candidate_id: str
    job_id: str
    resume_id: Optional[str] = None
    status: Literal[
        "Applied", "Screening", "Shortlisted", "Interview",
        "Technical", "HR", "Offer", "Hired", "Rejected"
    ] = "Applied"
    source: Optional[str] = "resume_upload"
    notes: Optional[str] = None


class ApplicationUpdate(BaseModel):
    status: Optional[Literal[
        "Applied", "Screening", "Shortlisted", "Interview",
        "Technical", "HR", "Offer", "Hired", "Rejected"
    ]] = None
    pipeline_order: Optional[int] = None
    source: Optional[str] = None
    notes: Optional[str] = None


class CandidateCreate(BaseModel):
    first_name: str
    last_name: Optional[str] = ""
    email: str
    phone: Optional[str] = None
    address: Optional[str] = None
    linkedin: Optional[str] = None
    github: Optional[str] = None
    portfolio: Optional[str] = None
    headline: Optional[str] = None
    current_company: Optional[str] = None
    current_designation: Optional[str] = None
    total_experience: Optional[str] = None
    relevant_experience: Optional[str] = None
    current_package: Optional[str] = None
    expected_package: Optional[str] = None
    notice_period: Optional[str] = None
    preferred_location: Optional[str] = None
    employment_type: Optional[str] = None
    summary: Optional[str] = None


class CandidateUpdate(BaseModel):
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    address: Optional[str] = None
    linkedin: Optional[str] = None
    github: Optional[str] = None
    portfolio: Optional[str] = None
    headline: Optional[str] = None
    current_company: Optional[str] = None
    current_designation: Optional[str] = None
    total_experience: Optional[str] = None
    relevant_experience: Optional[str] = None
    current_package: Optional[str] = None
    expected_package: Optional[str] = None
    notice_period: Optional[str] = None
    preferred_location: Optional[str] = None
    employment_type: Optional[str] = None
    summary: Optional[str] = None


class CandidateCompensationUpdate(BaseModel):
    current_package: Optional[str] = None
    expected_package: Optional[str] = None
    notice_period: Optional[str] = None


# ── Comments ──────────────────────────────────────────────────────────────────

class CommentCreate(BaseModel):
    body: str
    mentions: List[str] = []  # list of user_ids


class CommentUpdate(BaseModel):
    body: str
    mentions: List[str] = []


class CommentOut(BaseModel):
    id: str
    application_id: str
    author_id: str
    author_name: str
    body: str
    mentions: List[str] = []
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ── Activity ──────────────────────────────────────────────────────────────────

class ActivityOut(BaseModel):
    id: str
    action: str
    details: Optional[str] = None
    author_name: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Kanban card ───────────────────────────────────────────────────────────────

class NotificationOut(BaseModel):
    id: str
    title: str
    message: str
    link: Optional[str] = None
    is_read: bool
    level: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CandidateSnippet(BaseModel):
    id: str
    first_name: str
    last_name: str
    email: str
    headline: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class ApplicationKanbanOut(BaseModel):
    id: str
    status: str
    pipeline_order: int
    match_score: Optional[float] = None
    match_confidence: Optional[float] = None
    shortlist_reason: Optional[str] = None
    source: Optional[str] = None
    notes: Optional[str] = None
    applied_at: datetime
    updated_at: datetime
    candidate: CandidateSnippet
    comment_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class KanbanBoardOut(BaseModel):
    job_id: str
    job_title: str
    columns: dict[str, List[ApplicationKanbanOut]]


class InterviewCreate(BaseModel):
    candidate_id: str
    job_id: str
    application_id: Optional[str] = None
    interview_type: Optional[Literal[
        "HR Interview", "Technical Interview", "Manager Round", "Final Round", "Client Round"
    ]] = None
    scheduled_at: datetime
    duration_minutes: Optional[int] = None
    duration_minutes_str: Optional[str] = None
    time_zone: Optional[str] = None
    meeting_link: Optional[str] = None
    office_location: Optional[str] = None
    location: Optional[str] = None
    mode: Optional[str] = None
    interviewer: Optional[str] = None
    interviewer_user_id: Optional[str] = None
    notes: Optional[str] = None


class InterviewUpdate(BaseModel):
    interview_type: Optional[Literal[
        "HR Interview", "Technical Interview", "Manager Round", "Final Round", "Client Round"
    ]] = None
    scheduled_at: Optional[datetime] = None
    duration_minutes: Optional[int] = None
    duration_minutes_str: Optional[str] = None
    time_zone: Optional[str] = None
    meeting_link: Optional[str] = None
    office_location: Optional[str] = None
    location: Optional[str] = None
    mode: Optional[str] = None
    interviewer: Optional[str] = None
    interviewer_user_id: Optional[str] = None
    status: Optional[Literal[
        "Scheduled", "Completed", "Cancelled", "No Show", "Rescheduled"
    ]] = None
    notes: Optional[str] = None


class InterviewOut(BaseModel):
    id: str
    application_id: Optional[str] = None
    candidate_id: str
    candidate_name: Optional[str] = None
    job_id: str
    job_title: Optional[str] = None
    interview_type: Optional[str] = None
    scheduled_at: datetime
    duration_minutes: Optional[int] = None
    duration_minutes_str: Optional[str] = None
    time_zone: Optional[str] = None
    meeting_link: Optional[str] = None
    office_location: Optional[str] = None
    location: Optional[str] = None
    mode: Optional[str] = None
    interviewer: Optional[str] = None
    interviewer_user_id: Optional[str] = None
    interviewer_name: Optional[str] = None
    created_by_id: Optional[str] = None
    created_by_name: Optional[str] = None
    status: str
    rescheduled_from_id: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/pipeline.ts
// Pipeline types for the Kanban recruitment board

export const PIPELINE_STAGES = [
  "Applied",
  "Screening",
  "Shortlisted",
  "Interview",
  "Technical",
  "HR",
  "Offer",
  "Hired",
  "Rejected",
] as const;

export type PipelineStage = (typeof PIPELINE_STAGES)[number];

export interface CandidateSnippet {
  id: string;
  first_name: string;
  last_name: string;
  email: string;
  headline?: string | null;
}

export interface ApplicationCard {
  id: string;
  status: PipelineStage;
  pipeline_order: number;
  match_score?: number | null;
  match_confidence?: number | null;
  shortlist_reason?: string | null;
  source?: string | null;
  notes?: string | null;
  applied_at: string;
  updated_at: string;
  candidate: CandidateSnippet;
  comment_count: number;
}

export interface KanbanBoardData {
  job_id: string;
  job_title: string;
  columns: Record<PipelineStage, ApplicationCard[]>;
}

export interface Comment {
  id: string;
  application_id: string;
  author_id: string;
  author_name: string;
  body: string;
  mentions: string[];
  created_at: string;
}

export interface Activity {
  id: string;
  action: string;
  details?: string | null;
  author_name?: string | null;
  created_at: string;
}

export interface MentionUser {
  id: string;
  full_name: string;
  email: string;
}

// Stage display config
export const STAGE_CONFIG: Record<
  PipelineStage,
  { label: string; color: string; bgGradient: string; dot: string; textColor: string }
> = {
  Applied:     { label: "Applied",     color: "border-slate-500",   bgGradient: "from-slate-800 to-slate-900",     dot: "bg-slate-400",    textColor: "text-slate-300" },
  Screening:   { label: "Screening",   color: "border-blue-500",    bgGradient: "from-blue-900/60 to-slate-900",   dot: "bg-blue-400",     textColor: "text-blue-300"  },
  Shortlisted: { label: "Shortlisted", color: "border-violet-500",  bgGradient: "from-violet-900/60 to-slate-900", dot: "bg-violet-400",   textColor: "text-violet-300"},
  Interview:   { label: "Interview",   color: "border-amber-500",   bgGradient: "from-amber-900/60 to-slate-900",  dot: "bg-amber-400",    textColor: "text-amber-300" },
  Technical:   { label: "Technical",   color: "border-orange-500",  bgGradient: "from-orange-900/60 to-slate-900", dot: "bg-orange-400",   textColor: "text-orange-300"},
  HR:          { label: "HR Round",    color: "border-pink-500",    bgGradient: "from-pink-900/60 to-slate-900",   dot: "bg-pink-400",     textColor: "text-pink-300"  },
  Offer:       { label: "Offer",       color: "border-cyan-500",    bgGradient: "from-cyan-900/60 to-slate-900",   dot: "bg-cyan-400",     textColor: "text-cyan-300"  },
  Hired:       { label: "Hired ✓",    color: "border-emerald-500", bgGradient: "from-emerald-900/60 to-slate-900",dot: "bg-emerald-400",  textColor: "text-emerald-300"},
  Rejected:    { label: "Rejected",    color: "border-rose-500",    bgGradient: "from-rose-900/60 to-slate-900",   dot: "bg-rose-400",     textColor: "text-rose-300"  },
};
